File: tests/test_backtester.cpp
#include <doctest.h>
#include "../include/loader.h"
#include "../include/backtester.h"
using namespace std;
using namespace backtester;

TEST_CASE("Test strategy results")
{
    Dataset dataset = Loader::LoadDataset("../dataset");
    Evaluator::SetStrategyEvaluatorDataset(dataset);

    const string strategyProgram = R"(Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time))";
    vector<bool> strategyResults = Evaluator::RunStrategy(strategyProgram, "AAPL");

    CHECK((strategyResults.at(13) == false));
    CHECK((strategyResults.at(14) == true));
}

TEST_CASE("Test backtester")
{
    Dataset dataset = Loader::LoadDataset("../dataset");
    Evaluator::SetStrategyEvaluatorDataset(dataset);

    const string strategyProgram = R"(Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time))";
    CHECK((Evaluator::ValidateStrategyProgram(strategyProgram).first == true));

    vector<bool> strategyResults = Evaluator::RunStrategy(strategyProgram, "AAPL");
    vector<ExecutionData> executionData = Backtester::BacktestStoplossProfittake(strategyResults, "AAPL",
                                                                                 0.05, 0.05, 0.05);

    vector<double> returns = Returns::SimpleReturns(executionData, 0.05);
    double total = VectorOps::Total(returns);

    CHECK((abs(total) > 0));

}

TEST_CASE("Test backtester ind quantile")
{
    Dataset dataset = Loader::LoadDataset("../dataset");
    Evaluator::SetStrategyEvaluatorDataset(dataset);

    const string strategyProgram = R"(Indicator("EMA", stock, time) < IndQuantile("ClosePrice", "0.75",  stock, time))";
    CHECK((Evaluator::ValidateStrategyProgram(strategyProgram).first == true));

    vector<bool> strategyResults = Evaluator::RunStrategy(strategyProgram, "AAPL");
    vector<ExecutionData> executionData = Backtester::BacktestStoplossProfittake(strategyResults, "AAPL",
                                                                                 0.05, 0.05, 0.05);

    vector<double> returns = Returns::SimpleReturns(executionData, 0.05);
    double total = VectorOps::Total(returns);

    CHECK((abs(total) > 0));
}

TEST_CASE("Test parallel backtester")
{
    Dataset dataset = Loader::LoadDataset("../dataset");
    Evaluator::SetStrategyEvaluatorDataset(dataset);

    const string strategyProgram = R"(Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time))";
    CHECK((Evaluator::ValidateStrategyProgram(strategyProgram).first == true));

    map<string, vector<bool>> strategyResults = Evaluator::RunStrategyAllStocks(strategyProgram);

    vector<double> returns;
    for (auto const& [key, val] : strategyResults)
    {
        vector<ExecutionData> executionData = Backtester::BacktestStoplossProfittake(val, key,
                                                                                     0.05, 0.05, 0.05);

        double ret = VectorOps::Total(Returns::SimpleReturns(executionData, 0.05));
        returns.push_back(ret);
        cout << "Stock: " << key << ", Returns: " << ret << endl;
    }

    double totalRet = VectorOps::Total(returns);
    CHECK((abs(totalRet) > 0));
}

File: python/tests/test_backtester.py
import unittest
import os
from TradingStrategyBacktester.TradingStrategyBacktester import *


class TestBacktester(unittest.TestCase):
    def test_strategy_results(self):
        path = os.path.join(os.getcwd(), "..", "..", "dataset")
        dataset: dict[str, StockData] = Loader.LoadDataset(path)
        Evaluator.SetStrategyEvaluatorDataset(dataset)

        strategy_program: str = 'Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time)'
        strategy_results: list[bool] = Evaluator.RunStrategy(strategy_program, "AAPL")

        self.assertEqual(strategy_results[13], False)
        self.assertEqual(strategy_results[14], True)

    def test_backtester(self):
        path = os.path.join(os.getcwd(), "..", "..", "dataset")
        dataset: dict[str, StockData] = Loader.LoadDataset(path)
        Evaluator.SetStrategyEvaluatorDataset(dataset)

        strategy_program: str = 'Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time)'
        self.assertTrue(Evaluator.ValidateStrategyProgram(strategy_program))

        strategy_results: list[bool] = Evaluator.RunStrategy(strategy_program, "AAPL")

        execution_data: list[ExecutionData] = Backtester.BacktestStoplossProfittake(strategy_results, "AAPL",
                                                                                    0.05, 0.05, 0.05, -1)
        returns: list[float] = Returns.SimpleReturns(execution_data, 0.05)

        self.assertGreater(abs(sum(returns)), 0)

    def test_backtester_ind_quantile(self):
        path = os.path.join(os.getcwd(), "..", "..", "dataset")
        dataset: dict[str, StockData] = Loader.LoadDataset(path)
        Evaluator.SetStrategyEvaluatorDataset(dataset)

        strategy_program: str = 'Indicator("EMA", stock, time) < IndQuantile("ClosePrice", "0.75",  stock, time)'
        self.assertTrue(Evaluator.ValidateStrategyProgram(strategy_program))

        strategy_results: list[bool] = Evaluator.RunStrategy(strategy_program, "AAPL")

        execution_data: list[ExecutionData] = Backtester.BacktestStoplossProfittake(strategy_results, "AAPL",
                                                                                    0.05, 0.05, 0.05, -1)
        returns: list[float] = Returns.SimpleReturns(execution_data, 0.05)

        self.assertGreater(abs(sum(returns)), 0)

    def test_parallel_backtester(self):
        path = os.path.join(os.getcwd(), "..", "..", "dataset")
        dataset: dict[str, StockData] = Loader.LoadDataset(path)
        Evaluator.SetStrategyEvaluatorDataset(dataset)

        strategy_program: str = 'Indicator("EMA", stock, time) < Indicator("ClosePrice", stock, time)'
        strategy_results: dict[str, list[bool]] = Evaluator.RunStrategyAllStocks(strategy_program)

        returns = []
        for stock in strategy_results:
            execution_data: list[ExecutionData] = Backtester.BacktestStoplossProfittake(strategy_results.get(stock),
                                                                                        stock, 0.05, 0.05, 0.05, -1)
            ret = sum(Returns.SimpleReturns(execution_data, 0.05))
            print("Stock: " + stock + ", Return: " + str(ret))
            returns.append(ret)

        self.assertGreater(abs(sum(returns)), 0)



if __name__ == '__main__':
    unittest.main()
